File: 0502-ipo/0502-ipo.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        n = len(profits)
        projects = [(capital[i], profits[i]) for i in range(n)]
        projects.sort()
        maxheap = []
        i = 0
        for _ in range(k):
            # 檢查專案是否能夠以當前資本進行，如果可以，則將其可能的利潤（負值）加入最大堆 maxheap 中，並將索引 i 增加
            while i < n and projects[i][0] <= w: 
                heapq.heappush(maxheap, -projects[i][1])
                i += 1
            if not maxheap:
                break
            w -= heapq.heappop(maxheap)
        
        return w

File: 0502-ipo/0502-ipo.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        int n = profits.size();
        std::vector<std::pair<int, int>> projects;

        // Creating vector of pairs (capital, profits)
        for (int i = 0; i < n; i++){
            projects.emplace_back(capital[i], profits[i]);
        }

        // Sorting projects by capital required
        sort(projects.begin(), projects.end());

        // Max-heap to store profits, using greater to create a max-heap
        std::priority_queue<int> maxheap;
        int i = 0;

        // Main loop to select up to k projects
        for (int j = 0; j < k; j++){
            // Add all profitable projects that we can afford
            while (i < n && projects[i].first <= w){
                maxheap.push(projects[i].second);
                i++;
            }
            // If no projects can be funded, break out of the loop
            if (maxheap.empty()){
                break;
            }

            // Otherwise, take the project with the maximum profit
            w += maxheap.top(); // 取出 maxheap 中最大利潤的專案並增加當前資本 w。
            maxheap.pop();
        }
        return w;
    }
};